src/st/st-theme-node-drawing.c

```diff
@@ -590,18 +590,44 @@ create_cairo_pattern_of_background_gradient (StThemeNode *node,
     pattern = cairo_pattern_create_linear (0, 0, width, 0);
   else
     {
-      gdouble cx, cy;
+      gdouble cx, cy, radius;
 
-      cx = width / 2.;
-      cy = height / 2.;
-      pattern = cairo_pattern_create_radial (cx, cy, 0, cx, cy, MIN (cx, cy));
+      if (node->background_gradient_position_set)
+        {
+          cx = node->background_gradient_position_x;
+          cy = node->background_gradient_position_y;
+        }
+      else
+        {
+          cx = width / 2.;
+          cy = height / 2.;
+        }
+
+      if (node->background_gradient_radius != -1)
+        radius = node->background_gradient_radius;
+      else
+        radius = MIN (width / 2., height / 2.);
+
+      pattern = cairo_pattern_create_radial (cx, cy, 0, cx, cy, radius);
     }
 
   cairo_pattern_add_color_stop_rgba (pattern, 0,
                                      node->background_color.red / 255.,
                                      node->background_color.green / 255.,
                                      node->background_color.blue / 255.,
                                      node->background_color.alpha / 255.);
+
+  /* Add an intermediate stop when any is set */
```

Comment:
Is there a way you can generalize this to support more than one stop from the CSS?